runtime/src/pool.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::HashMap;
+
+use crate::{Error, FromBytesOwned, Plugin, PluginBuilder, ToBytes};
+
+/// `PoolPlugin` is used by the pool to track the number of live instances of a particular plugin
+#[derive(Clone, Debug)]
+pub struct PoolPlugin(std::rc::Rc<std::cell::RefCell<Plugin>>);
+
+impl PoolPlugin {
+    fn new(plugin: Plugin) -> Self {
+        Self(std::rc::Rc::new(std::cell::RefCell::new(plugin)))
+    }
+
+    /// Access the underlying plugin
+    pub fn plugin(&self) -> std::cell::RefMut<Plugin> {
+        self.0.borrow_mut()
+    }
+
+    /// Helper to call a plugin function on the underlying plugin
+    pub fn call<'a, Input: ToBytes<'a>, Output: FromBytesOwned>(
+        &self,
+        name: impl AsRef<str>,
+        input: Input,
+    ) -> Result<Output, Error> {
+        self.plugin().call(name.as_ref(), input)
+    }
+
+    /// Helper to get the underlying plugin's ID
+    pub fn id(&self) -> uuid::Uuid {
+        self.plugin().id
+    }
+}
+
+type PluginSource = dyn Fn() -> Result<Plugin, Error>;
+
+struct PoolInner<Key: std::fmt::Debug + Clone + std::hash::Hash + Eq = String> {
+    plugins: HashMap<Key, Box<PluginSource>>,
+    instances: HashMap<Key, Vec<PoolPlugin>>,
+}
+
+/// `Pool` manages threadsafe access to a limited number of instances of multiple plugins
+#[derive(Clone)]
+pub struct Pool<Key: std::fmt::Debug + Clone + std::hash::Hash + Eq = String> {
```

Comment:
Im not sure what the `Key` parameter is good for. It may be useful to store multiple different plugin types in the same pool, but this is tricky as there is currently no way to get a list of all existing keys. Anyway the pool itself has very little overhead, so it makes more sense for users to write `HashMap<Key, Pool>` if needed. In my case Im using `Vec<Pool<()>>` because there is no need for any hashmap.